src/error_handling.c

```diff
@@ -14,313 +14,161 @@
+  // Only warn of overwritting if the new error is different from the old ones.
+  size_t characters_to_compare = strnlen(error_string, RCUTILS_ERROR_MESSAGE_MAX_LENGTH);
+  // assumption is that message length is <= max error string length
+  static_assert(
+    sizeof(gtls_rcutils_error_state.message) <= sizeof(gtls_rcutils_error_string.str),
+    "expected error state's max message length to be less than or equal to error string max");
+  if (
+    gtls_rcutils_error_is_set &&
+    __same_string(error_string, gtls_rcutils_error_string.str, characters_to_compare) &&
+    __same_string(error_string, gtls_rcutils_error_state.message, characters_to_compare))
+  {
+    fprintf(
+      stderr,
+      "\n"
+      ">>> [rcutils|error_handling.c:" RCUTILS_STRINGIFY(__LINE__) "] rcutils_set_error_state()\n"
+      "This error state is being overwritten:\n"
+      "\n"
+      "  '%s'\n"
+      "\n"
+      "with this new error message:\n"
+      "\n"
+      "  '%s, at %s:%zu'\n"
+      "\n"
+      "rcutils_reset_error() should be called after error handling to avoid this.\n"
+      "<<<\n",
+      gtls_rcutils_error_string.str,
+      error_string,
+      file,
+      line_number);
+  }
+#endif
+  gtls_rcutils_error_state = error_state;
+  gtls_rcutils_error_string_is_formatted = false;
+  gtls_rcutils_error_string = (const rcutils_error_string_t) {
+    .str = "\0"
+  };
+  gtls_rcutils_error_is_set = true;
 }
 
 bool
 rcutils_error_is_set(void)
 {
-#ifdef RCUTILS_THREAD_LOCAL_PTHREAD
-  rcutils_error_state_t * __rcutils_error_state =
-    (rcutils_error_state_t *)pthread_getspecific(__rcutils_error_state_key);
-#endif
-  return __rcutils_error_is_set(__rcutils_error_state);
+  return gtls_rcutils_error_is_set;
 }
 
-const char *
-rcutils_get_error_string_safe(void)
+const rcutils_error_state_t *
+rcutils_get_error_state(void)
 {
-  if (!rcutils_error_is_set()) {
-    return "error not set";
-  }
-  return rcutils_get_error_string();
+  return &gtls_rcutils_error_state;
 }
 
-void
-__rcutils_reset_error_string(char ** error_string_ptr, rcutils_allocator_t allocator)
+rcutils_error_string_t
+rcutils_get_error_string(void)
 {
-  if (NULL == error_string_ptr) {
-    return;
-  }
-
-  rcutils_allocator_t local_allocator = allocator;
-  if (!rcutils_allocator_is_valid(&allocator)) {
-#if RCUTILS_REPORT_ERROR_HANDLING_ERRORS
-    RCUTILS_SAFE_FWRITE_TO_STDERR(
-      "[rcutils|error_handling.c:" RCUTILS_STRINGIFY(__LINE__) "]: "
-      "invalid allocator\n");
-#endif
-    local_allocator = rcutils_get_default_allocator();
+  if (!gtls_rcutils_error_is_set) {
+    return (rcutils_error_string_t) {"error not set"};  // NOLINT(readability/braces)
   }
-  char * error_string = *error_string_ptr;
-  if (error_string != NULL) {
-    local_allocator.deallocate(error_string, local_allocator.state);
+  if (!gtls_rcutils_error_string_is_formatted) {
+    __rcutils_format_error_string(&gtls_rcutils_error_string, &gtls_rcutils_error_state);
```

Comment:
Is `gtls_rcutils_error_string_is_formatted` supposed to be set to `true` after this?

Reply:
Yup! Good catch. I don't have a good way to test this atm.